hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -0,0 +1,402 @@
+		if nodePool.Annotations[nodePoolAnnotationCurrentConfig] != targetConfigHash {
+			log.Info("Config update complete",
+				"previous", nodePool.Annotations[nodePoolAnnotationCurrentConfig], "new", targetConfigHash)
+			nodePool.Annotations[nodePoolAnnotationCurrentConfig] = targetConfigHash
+		}
+		nodePool.Annotations[nodePoolAnnotationCurrentConfigVersion] = targetConfigVersionHash
+		return nil
+	}
+
+	// Otherwise:
+	// Order Nodes deterministically.
+	// Check state: AtVersionConfig, Upgrading, wantVersionConfig.
+	// If AtVersionConfig then next Node.
+	// If Upgrading then no-op, return.
+	// If wantVersionConfig then:
+	// Check maxUnavailable/MaxSurge.
+	// Drain.
+	// Create Namespace/RBAC/ConfigMap/Pod in guest cluster.
+	// Mark Node as Upgrading.
+	return nil
+}
+
+func (r *NodePoolReconciler) nodeToNodePool(o client.Object) []reconcile.Request {
+	node, ok := o.(*corev1.Node)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Node but got a %T", o))
+	}
+
+	machineName, ok := node.GetAnnotations()[capiv1.MachineAnnotation]
+	if !ok {
+		return nil
+	}
+
+	// Match by namespace when the node has the annotation.
+	machineNamespace, ok := node.GetAnnotations()[capiv1.ClusterNamespaceAnnotation]
+	if !ok {
+		return nil
+	}
+
+	// Match by nodeName and status.nodeRef.name.
+	machine := &capiv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: machineNamespace,
+			Name:      machineName,
+		},
+	}
+	if err := r.Client.Get(context.TODO(), client.ObjectKeyFromObject(machine), machine); err != nil {
+		return nil
+	}
+
+	machineOwner := metav1.GetControllerOf(machine)
+	if machineOwner.Kind != "MachineSet" {
+		return nil
+	}
+
+	machineSet := &capiv1.MachineSet{ObjectMeta: metav1.ObjectMeta{
+		Name:      machineOwner.Name,
+		Namespace: machineNamespace,
+	}}
+	if err := r.Client.Get(context.TODO(), client.ObjectKeyFromObject(machineSet), machineSet); err != nil {
+		return nil
+	}
+
+	nodePoolName := machineSet.GetAnnotations()[nodePoolAnnotation]
+	if nodePoolName == "" {
+		return nil
+	}
+
+	return []reconcile.Request{
+		{NamespacedName: hyperutil.ParseNamespacedName(nodePoolName)},
+	}
+}
+
+func getNodesForMachineSet(ctx context.Context, c client.Reader, hostedClusterClient client.Client, machineSet *capiv1.MachineSet) ([]*corev1.Node, error) {
+	selectorMap, err := metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert MachineSet %q label selector to a map: %v", machineSet.Name, err)
+	}
+
+	// Get all Machines linked to this MachineSet.
+	allMachines := &capiv1.MachineList{}
+	if err = c.List(ctx,
+		allMachines,
+		client.InNamespace(machineSet.Namespace),
+		client.MatchingLabels(selectorMap),
+	); err != nil {
+		return nil, fmt.Errorf("failed to list machines: %v", err)
+	}
+
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range allMachines.Items {
+		if metav1.GetControllerOf(&machine) != nil && metav1.IsControlledBy(&machine, machineSet) {
```

Comment:
do you need this additional check if you're already filtering by the machineset's label selector?

Reply:
yeh still needed for consistency, the machineSet controller must set it.